### PR TITLE
Replace craft preview transaction list with a single entry.

### LINKED ISSUE #2505: Issue with the `CraftItemEvent.Preview`
**I am currently running**

- SpongeForge version: spongeforge-1.12.2-2838-7.1.10-RC3994
- SpongeVanilla version: 
- Java version: 1.8.0_231
- Operating System: windows 10

- Plugins/Mods: only a test plugin with a listener
```java
    @Listener
    public void onCraft(CraftItemEvent.Preview event) {
        event.getTransactions().forEach(slotTransaction -> slotTransaction.setCustom(slotTransaction.getFinal()));
    }
```

**Issue Description**
When drag in crafting slots then release the left button, the output slot will show the craft as if there is only the first crafting slot is set.
images:
N4M`9Y%XGNZZ49AW{683PST
S})A$Q10UA6`IHL9_WOPTUJ
QTMSWXWFUBSS9KB8VL{~S)J

The event is fired form DragInventoryStopState
https://github.com/SpongePowered/SpongeCommon/blob/72a1d111b0aeceacf74f7595c26454f7e16a2737/src/main/java/org/spongepowered/common/event/tracking/phase/packet/drag/DragInventoryStopState.java#L61-L76

I add a breakpoint and find that the captured `previewTransactions` is more one.
In the first image, it is like:
1. air -> iron_nugget*9
2. iron_nugget*9 -> air
3. air -> air
4. air -> air

But we choose the first transaction as craft result.

### PR DESCRIPTION
**Pull request by JBYoshi**
Basically everything that uses the crafting transaction list relies on it having only a single entry. However, when the player drags in the crafting table, `slotChangedCraftingGrid` is called multiple times, putting multiple entries into the list. This PR merges these calls into a single transaction to prevent issues like this.

Fixes #2505.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/event/tracking/phase/packet/drag/DragInventoryStopState.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/event/tracking/phase/packet/drag/DragInventoryStopState.java]: (nitpick) `final` this value please.

File: src/main/java/org/spongepowered/common/event/tracking/phase/packet/inventory/PlaceRecipePacketState.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/event/tracking/phase/packet/inventory/PlaceRecipePacketState.java]: (nitpick) `final` this variable please

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java]: Could this be configurable?

File: src/main/java/org/spongepowered/common/mixin/core/inventory/SlotCraftingMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/SlotCraftingMixin.java]: (nitpick) `final` this variable please

**Review by gabizou — APPROVED** [review]
Nitpicks aside, looks good to me

**Comment by JBYoshi** [comment]
@gabizou Updated.

File: src/main/java/org/spongepowered/common/config/category/LoggingCategory.java
Comment (dualspiral) [inline src/main/java/org/spongepowered/common/config/category/LoggingCategory.java]: I know @gabizou said "could this be configurable?" and attached it to the log line, but I'm not sure this is what he meant - I think we should _always_ be reporting a failure to the log.

Was the meaning that this entire behaviour on failure should be configurable - as in, original vs replacement?

Comment (gabizou) [reply src/main/java/org/spongepowered/common/config/category/LoggingCategory.java]: Well this and whether silencing after X number of failed/merged transactions happens.

Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/config/category/LoggingCategory.java]: I can understand the configuration to "silence after X messages", but why would the behavior of the failure need to be configurable? I can't imagine a server owner changing that.

File: src/main/java/org/spongepowered/common/config/category/LoggingCategory.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/config/category/LoggingCategory.java]: Should this comment be multi-lined to roughly 100 characters?

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java:435]: Are you doing the correct equality check here? Or are you aiming to just allow instance equality checking the `ItemStackSnapshot`?

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java:434]: Off the top of my head, do `SlotAdapter`s implement `equals`?

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java]:
```suggestion
            final SlotTransaction replace = new SlotTransaction(slot, orig, repl);
```

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java]:
```suggestion
            final int maxLogs = SpongeImpl.getGlobalConfigAdapter().getConfig().getLogging().logTransactionMergeFailure();
```

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java]: Static fields should be in their own body imo. Make it well aware they are static.

**Review by gabizou — COMMENTED** [review]
Some nitpicks

**Comment by Zidane** [comment]
@JBYoshi Get this updated please.

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java:435]: It has to be `equals()` - `orig` is created fresh each time the method is called.

File: src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java
Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java:434]: Not currently. It seems to be an easy implementation - it just requires checking the underlying inventory and the index - so I'll do that.

**Comment by dualspiral** [comment]
@JBYoshi Is this good to go? Anything we need to be aware of if we merge it?

**Comment by JBYoshi** [comment]
I think it should be good to merge, if there aren't any other objections.

**Comment by dualspiral** [comment]
I'm sure someone will complain if it doesn't work.

**Pull request merged**
